yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/main.nr

```diff
@@ -0,0 +1,501 @@
+        }
+
+        let amount = SafeU120::new(amount);
+        let from_balance = storage.public_balances.at(from.address).read().sub(amount);
+        storage.public_balances.at(from.address).write(from_balance);
+
+        let new_supply = storage.total_supply.read().sub(amount);
+        storage.total_supply.write(new_supply);
+
+        1
+    }
+    // docs:end:burn_public
+
+    // docs:start:redeem_shield
+    #[aztec(private)]
+    fn redeem_shield(
+        to: AztecAddress,
+        amount: Field,
+        secret: Field,
+    ) -> Field {
+        let slow = SlowMap::at(AztecAddress::new(storage.slow_update.get_note().value));        
+        let to_roles = UserFlags::new(slow.read_at(&mut context, to.address) as u120);
+        assert(!to_roles.is_blacklisted, "Blacklisted: Recipient");
+
+        let pending_shields = storage.pending_shields;
+        let secret_hash = compute_secret_hash(secret);
+        // Get 1 note (set_limit(1)) which has amount stored in field with index 0 (select(0, amount)) and secret_hash
+        // stored in field with index 1 (select(1, secret_hash)).
+        let options = NoteGetterOptions::new().select(0, amount).select(1, secret_hash).set_limit(1);
+        let notes = pending_shields.get_notes(options);
+        let note = notes[0].unwrap_unchecked();
+        // Remove the note from the pending shields set 
+        pending_shields.remove(note);
+
+        // Add the token note to user's balances set
+        storage.balances.at(to).add(SafeU120::new(amount));
+
+        1
+    }
+    // docs:end:redeem_shield
+
+    // docs:start:unshield
+    #[aztec(private)]
+    fn unshield(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let slow = SlowMap::at(AztecAddress::new(storage.slow_update.get_note().value));
+        let from_roles = UserFlags::new(slow.read_at(&mut context, from.address) as u120);
+        assert(!from_roles.is_blacklisted, "Blacklisted: Sender");
+        let to_roles = UserFlags::new(slow.read_at(&mut context, to.address) as u120);
+        assert(!to_roles.is_blacklisted, "Blacklisted: Recipient");
+
+        if (from.address != context.msg_sender()) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        storage.balances.at(from).sub(SafeU120::new(amount));
+
+        let selector = compute_selector("_increase_public_balance((Field),Field)");
+        let _void = context.call_public_function(context.this_address(), selector, [to.address, amount]);
+
+        1
+    }
+    // docs:end:unshield
+
+    // docs:start:transfer
+    #[aztec(private)]
+    fn transfer(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let slow = SlowMap::at(AztecAddress::new(storage.slow_update.get_note().value));
```

Comment:
I think it would be clearer if the interface was called SlowTree given that it is calling the slow tree contract. It confused me for a bit when I instinctively navigated to read_at functionality in SlowMap struct and I was confused why public context was used there (given that bellow it's calling a private func).